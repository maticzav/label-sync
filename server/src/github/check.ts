/**
 * This file contains utility functions that we use to communicate
 * with GitHub about issues and related content.
 */

import { ProbotOctokit } from 'probot'
import * as t from 'luxon'

type Octokit = InstanceType<typeof ProbotOctokit>

// MARK: - Methods

export type CheckResult = {
  success: boolean
  /**
   * The title of the check.
   */
  title: string
  /**
   * A brief summary.
   */
  summary: string
  /**
   * Link to details page.
   */
  details_url?: string
}

type CheckParams = {
  owner: string
  repo: string
  /**
   * Name of the check event.
   */
  name: string
  /**
   * SHA of the head.
   */
  sha: string
}

/**
 * Creates a check and executes a handler function that resolves
 * with a check result.
 */
export async function check(
  github: Octokit,
  params: CheckParams,
  handler: () => Promise<CheckResult>,
) {
  // https://docs.github.com/en/rest/reference/checks#update-a-check-run

  /**
   * This function creates a new check, executes a handler
   * and updates check result with the event output.
   */

  const check = await github.checks
    .create({
      name: 'label-sync/dryrun',
      owner: params.owner,
      repo: params.repo,
      head_sha: params.sha,
      started_at: t.DateTime.now().toISO(),
      status: 'in_progress',
    })
    .then((res) => res.data)

  // Execute
  let result: CheckResult

  try {
    /**
     * Executes the handler in a wrapped environment.
     */
    result = await handler()
  } catch (err) /* instabul ignore next */ {
    const completion = await github.checks.update({
      check_run_id: check.id,
      owner: params.owner,
      repo: params.repo,
      status: 'completed',
      completed_at: t.DateTime.now().toISO(),
      conclusion: 'failure',
    })

    return completion.data
  }

  // Complete

  const completion = await github.checks.update({
    check_run_id: check.id,
    owner: params.owner,
    repo: params.repo,
    status: 'completed',
    completed_at: t.DateTime.now().toISO(),
    conclusion: result.success ? 'succeess' : 'failure',
    details_url: result.details_url,
    output: {
      title: result.title,
      summary: result.summary,
    },
  })

  return completion.data
}
