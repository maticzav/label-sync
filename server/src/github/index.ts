export * from './check'
export * from './issues'
export * from './labels'
export * from './repository'
export * from './tree'
export * from './pullrequest'
