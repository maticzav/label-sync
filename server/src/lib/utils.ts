/**
 * Negates the wrapped function.
 * @param fn
 */
export function not<T, L extends Array<T>>(fn: (...args: L) => boolean): (...args: L) => boolean {
  return (...args) => !fn(...args)
}

/**
 * Creates a fallback default value.
 * @param fallback
 * @param value
 */
export function withDefault<T>(fallback: T, value: T | undefined | null): T {
  if (value == null) return fallback
  return value
}
