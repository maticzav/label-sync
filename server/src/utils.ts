import fs from 'fs'
import path from 'path'
import { mapKeys } from './data/dict'

/**
 * Negates the wrapped function.
 * @param fn
 */
export function not<T, L extends Array<T>>(
  fn: (...args: L) => boolean,
): (...args: L) => boolean {
  return (...args) => !fn(...args)
}

/**
 * Loads a tree of utf-8 decoded files at paths.
 *
 * @param path
 */
export function loadTreeFromPath(
  root: string,
  ignore: (string | RegExp)[],
): { [path: string]: string } {
  const files = fs.readdirSync(root, { encoding: 'utf-8' })
  const tree = files
    .filter(file => !ignore.some(glob => RegExp(glob).test(file)))
    .flatMap(file => {
      const rootFilePath = path.resolve(root, file)
      if (fs.lstatSync(rootFilePath).isDirectory()) {
        return Object.entries(
          mapKeys(loadTreeFromPath(rootFilePath, ignore), key =>
            unshift(file, key),
          ),
        )
      } else {
        return [[file, fs.readFileSync(rootFilePath, { encoding: 'utf-8' })]]
      }
    })

  return Object.fromEntries(tree)
}

/**
 * Adds a folder to the path.
 * @param pre
 * @param path
 */
function unshift(pre: string, path: string): string {
  return [pre, ...path.split('/').filter(Boolean)].join('/')
}
