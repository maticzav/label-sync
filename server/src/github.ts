import * as path from 'path'
import { Octokit } from 'probot'

import { Maybe } from './data/maybe'
import { Dict, mapEntriesAsync } from './data/dict'
import { not } from './utils'

/**
 * Loads a file from Github.
 *
 * @param octokit
 * @param path
 */
export async function getFile(
  octokit: Octokit,
  { owner, repo, ref }: { owner: string; repo: string; ref: string },
  path: string,
): Promise<Maybe<string>> {
  try {
    const res = await octokit.repos.getContents({
      owner: owner,
      path: path,
      repo: repo,
      ref: ref,
    })

    switch (res.status) {
      case 200: {
        // expect a single file
        /* istanbul ignore if */
        if (Array.isArray(res.data) || !res.data.content) return null

        return Buffer.from(res.data.content, 'base64').toString()
      }
      /* istanbul ignore next */
      default: {
        return null
      }
    }
  } catch (err) /* istanbul ignore next */ {
    return null
  }
}

export interface GithubLabel {
  name: string
  description?: string
  color: string
  default?: boolean
}

// export interface GithubLabelDiff extends GithubLabel {
//   old: Partial<GithubLabel>
// }

/**
 * Fetches labels in a repository.
 */
export async function getRepositoryLabels(
  octokit: Octokit,
  { repo, owner }: { repo: string; owner: string },
): Promise<Maybe<Octokit.IssuesListLabelsForRepoResponseItem[]>> {
  try {
    const labels = octokit.issues.listLabelsForRepo.endpoint.merge({
      repo,
      owner,
    })

    return octokit.paginate(labels)
  } catch (err) /* istanbul ignore next */ {
    return null
  }
}

/**
 *
 * Create new labels in a repository.
 *
 * @param github
 * @param labels
 * @param repository
 */
export async function addLabelsToRepository(
  github: Octokit,
  { repo, owner }: { repo: string; owner: string },
  labels: GithubLabel[],
  persist: boolean,
): Promise<GithubLabel[]> {
  /* Return immediately on non-persistent sync. */
  if (!persist) return labels

  /* Perform sync on persist. */
  const actions = labels.map(label => addLabelToRepository(label))
  await Promise.all(actions)

  return labels

  /**
   * Helper functions
   */
  async function addLabelToRepository(
    label: GithubLabel,
  ): Promise<GithubLabel> {
    return github.issues
      .createLabel({
        owner: owner,
        repo: repo,
        name: label.name,
        description: label.description,
        color: label.color,
      })
      .then(res => res.data)
  }
}

/**
 *
 * Updates labels in repository.
 *
 * @param github
 * @param labels
 * @param repository
 */
export async function updateLabelsInRepository(
  github: Octokit,
  { repo, owner }: { repo: string; owner: string },
  labels: GithubLabel[],
  persist: boolean,
): Promise<GithubLabel[]> {
  /* Return immediately on non-persistent sync. */
  if (!persist) return labels

  /* Update values on persist. */
  const actions = labels.map(label => updateLabelInRepository(label))
  await Promise.all(actions)

  return labels

  /**
   * Helper functions
   */
  async function updateLabelInRepository(
    label: GithubLabel,
  ): Promise<GithubLabel> {
    return github.issues
      .updateLabel({
        current_name: label.name,
        owner: owner,
        repo: repo,
        name: label.name,
        description: label.description,
        color: label.color,
      })
      .then(res => res.data)
  }
}

/**
 *
 * Removes labels from repository.
 *
 * @param github
 * @param labels
 * @param repository
 */
export async function removeLabelsFromRepository(
  github: Octokit,
  { repo, owner }: { repo: string; owner: string },
  labels: GithubLabel[],
  persist: boolean,
): Promise<GithubLabel[]> {
  /* Return immediately on non-persistent sync. */
  if (!persist) return labels

  const actions = labels.map(label => removeLabelFromRepository(label))
  await Promise.all(actions)

  return labels

  /**
   * Helper functions
   */
  async function removeLabelFromRepository(
    label: GithubLabel,
  ): Promise<Octokit.IssuesDeleteLabelParams> {
    return github.issues
      .deleteLabel({
        owner: owner,
        repo: repo,
        name: label.name,
      })
      .then(res => res.data)
  }
}

/**
 *
 * Compares two labels by comparing all of their keys.
 *
 * @param label
 */
export function isLabel(local: GithubLabel): (remote: GithubLabel) => boolean {
  return remote =>
    local.name === remote.name &&
    local.description === remote.description &&
    local.color === remote.color
}

/**
 * Determines whether the two configurations configure the same label.
 *
 * @param local
 */
export function isLabelDefinition(
  local: GithubLabel,
): (remote: GithubLabel) => boolean {
  return remote => local.name === remote.name
}

/**
 * Opens an issue with a prescribed title and body.
 *
 * @param octokit
 * @param owner
 * @param reports
 */
export async function openIssue(
  octokit: Octokit,
  owner: string,
  repo: string,
  title: string,
  body: string,
): Promise<Octokit.IssuesCreateResponse> {
  return octokit.issues
    .create({
      repo: repo,
      owner: owner,
      title: title,
      body: body,
    })
    .then(({ data }) => data)
}

// /**
//  * Closes the issue openned by the LabelSync configuration.
//  *
//  * @param octokit
//  * @param owner
//  * @param repo
//  * @param title
//  */
// export async function closeIssue(
//   octokit: Octokit,
//   owner: string,
//   repo: string,
//   title: string,
// ) {
//   const issues = await octokit.issues.listForRepo({
//     owner: owner,
//     repo: repo,
//     creator: 'labelsync-manager',
//   })
// }

/**
 * Creates a comment on a dedicated pull request.
 * @param octokit
 * @param owner
 * @param repo
 * @param number
 * @param message
 */
export async function createPRComment(
  octokit: Octokit,
  owner: string,
  repo: string,
  number: number,
  message: string,
): Promise<Octokit.IssuesCreateCommentResponse> {
  return octokit.issues
    .createComment({
      owner: owner,
      repo: repo,
      body: message,
      issue_number: number,
    })
    .then(({ data }) => data)
}

/**
 * Tries to fetch a repository.
 *
 * @param github
 * @param owner
 * @param repo
 */
export async function getRepo(
  github: Octokit,
  owner: string,
  repo: string,
): Promise<
  { status: 'Exists'; repo: Octokit.ReposGetResponse } | { status: 'Unknown' }
> {
  return github.repos
    .get({
      owner: owner,
      repo: repo,
    })
    .then(res => {
      switch (res.status) {
        case 200: {
          return { status: 'Exists' as const, repo: res.data }
        }
        /* istanbul ignore next */
        default: {
          return { status: 'Unknown' as const }
        }
      }
    })
    .catch(() => {
      return { status: 'Unknown' as const }
    })
}

type GHRepo = { owner: string; repo: string }

/**
 * Represents a Github file/folder structure.
 *
 * Files should be utf-8 strings.
 */
export type GHTree = { [path: string]: string }

/**
 * Returns the files that are not nested.
 * @param tree
 */
function getTreeFiles(tree: GHTree): Dict<string> {
  return Object.fromEntries(
    Object.keys(tree)
      .filter(isFileInThisFolder)
      .map(name => [name, tree[name]]),
  )
}

/**
 * Returns a dictionary of remaining subtrees.
 * @param tree
 */
function getTreeSubTrees(tree: GHTree): Dict<GHTree> {
  return Object.keys(tree)
    .filter(not(isFileInThisFolder))
    .reduce<Dict<GHTree>>((acc, filepath) => {
      const [subTree, newFilepath] = shiftPath(filepath)
      if (!acc.hasOwnProperty(subTree)) {
        acc[subTree] = {}
      }
      acc[subTree][newFilepath] = tree[filepath]
      return acc
    }, {})
}

/**
 * Shifts path by one.
 * Returns the shifted part as first argument and remaining part as second.
 *
 * @param filepath
 */
function shiftPath(filepath: string): [string, string] {
  const [dir, ...dirs] = filepath.split('/').filter(Boolean)
  return [dir, dirs.join('/')]
}

/**
 * Determines whether a path references a direct file
 * or a file in the nested folder.
 *
 * "/src/index.ts" -> false
 * "/index.ts" -> true
 * "index.ts" -> true
 *
 * @param filePath
 */
function isFileInThisFolder(filePath: string): boolean {
  return ['.', '/'].includes(path.dirname(filePath))
}

/**
 * Recursively creates a tree commit by creating blobs and generating
 * trees on folders.
 *
 * @param github
 * @param tree
 */
async function createGhTree(
  github: Octokit,
  { owner, repo }: GHRepo,
  tree: GHTree,
): Promise<{ sha: string }> {
  /**
   * Uploads blobs and creates subtrees.
   */
  const blobs = await mapEntriesAsync(getTreeFiles(tree), content =>
    createGhBlob(github, { owner, repo }, content),
  )
  const trees = await mapEntriesAsync(getTreeSubTrees(tree), subTree =>
    createGhTree(github, { owner, repo }, subTree),
  )

  return github.git
    .createTree({
      owner,
      repo,
      tree: [
        ...Object.entries(trees).map(([treePath, { sha }]) => ({
          mode: '040000' as const,
          type: 'tree' as const,
          path: treePath,
          sha,
        })),
        ...Object.entries(blobs).map(([filePath, { sha }]) => ({
          mode: '100644' as const,
          type: 'blob' as const,
          path: filePath,
          sha,
        })),
      ],
    })
    .then(res => res.data)
}

/**
 * Creates a Github Blob from a File.
 *
 * @param github
 * @param param1
 * @param file
 */
async function createGhBlob(
  github: Octokit,
  { owner, repo }: GHRepo,
  content: string,
): Promise<Octokit.GitCreateBlobResponse> {
  return github.git.createBlob({ owner, repo, content }).then(res => res.data)
}

/**
 * Bootstraps a configuration repository to a prescribed destination.
 *
 * Assumes repository is empty.
 *
 * @param github
 * @param owner
 * @param repo
 */
export async function bootstrapConfigRepository(
  github: Octokit,
  { owner, repo }: GHRepo,
  tree: GHTree,
): Promise<Octokit.GitUpdateRefResponse> {
  await github.repos
    .createInOrg({
      org: owner,
      name: repo,
      description: 'LabelSync configuration repository.',
      auto_init: true,
    })
    .then(res => res.data)

  const gitTree = await createGhTree(github, { owner, repo }, tree)
  const masterRef = await github.git
    .getRef({
      owner,
      repo,
      ref: 'heads/master',
    })
    .then(res => res.data)

  const commit = await github.git
    .createCommit({
      owner,
      repo,
      message: ':sparkles: Scaffold configuration',
      tree: gitTree.sha,
      parents: [masterRef.object.sha],
    })
    .then(res => res.data)

  const ref = await github.git.updateRef({
    owner,
    repo,
    ref: 'heads/master',
    sha: commit.sha,
  })

  return ref.data
}

export type InstallationAccess =
  | { status: 'Sufficient' }
  | { status: 'Insufficient'; missing: string[] }

/**
 * Determines whether LabelSync can access all requested repositories.
 * @param github
 * @param repos
 */
export async function checkInstallationAccess(
  github: Octokit,
  repos: string[],
): Promise<InstallationAccess> {
  const {
    data: { repositories },
  } = await github.apps.listRepos({ per_page: 100 })

  const missing = repos.filter(repo =>
    repositories.every(({ name }) => repo !== name),
  )

  if (missing.length === 0) {
    return { status: 'Sufficient' }
  }

  return {
    status: 'Insufficient',
    missing: missing,
  }
}
