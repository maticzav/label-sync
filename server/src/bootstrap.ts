import { compile } from 'handlebars'

import { mapEntries } from './data/dict'
import { GHTree } from './github'

/**
 * Populates the template with repositories.
 * @param tree
 */
export function populateTempalte(
  tree: GHTree,
  data: { repository: string; repositories: { name: string }[] },
): GHTree {
  return mapEntries(tree, (file) => compile(file)(data))
}
