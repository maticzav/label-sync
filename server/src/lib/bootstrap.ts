import * as handlebars from 'handlebars'
import * as path from 'path'
import * as prettier from 'prettier'

import { mapEntries } from '../data/dict'
import { GHTree } from './github'

/**
 * Populates the template with repositories.
 * @param tree
 */
export function populateTemplate(tree: GHTree, data: { repository: string; repositories: { name: string }[] }): GHTree {
  return mapEntries(tree, (file, name) => {
    /* Personalize file */
    const populatedFile = handlebars.compile(file)(data)

    /* Format it */
    switch (path.extname(name)) {
      case '.ts': {
        return prettier.format(populatedFile, { parser: 'typescript' })
      }
      case '.yml': {
        return prettier.format(populatedFile, { parser: 'yaml' })
      }
      case '.md': {
        return prettier.format(populatedFile, { parser: 'markdown' })
      }
      case '.json': {
        return prettier.format(populatedFile, { parser: 'json' })
      }
      default: {
        return populatedFile
      }
    }
  })
}
