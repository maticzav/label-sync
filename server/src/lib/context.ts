import _ from 'lodash'
import { ProbotOctokit } from 'probot'

import { getLSConfigRepoName, LSCConfiguration, LS_CONFIG_PATH, parseConfig } from '@labelsync/config'
import { Installation } from '@labelsync/database'

import { getFile } from './github'
import { Sources } from './sources'
import { DateTime } from 'luxon'
import pino from 'pino'

export type Octokit = InstanceType<typeof ProbotOctokit>

/**
 * Context that all events should include.
 */
interface BaseContext {
  octokit: Octokit
  payload: {
    action?: string | null
    repository: {
      name: string
      default_branch: string
      owner: {
        id: number
        login: string
        type: string
      }
    }
  }
  sources: Sources
}

/**
 * Looks up the installation associated with the account from a given event
 * and checks that the subscription is active.
 */
export async function getAccountInstallation(ctx: BaseContext): Promise<Installation> {
  const account = ctx.payload.repository.owner.login.toLowerCase()

  const installation = await ctx.sources.installations.get({ account })

  /* istanbul ignore if */
  if (installation == null) {
    throw new ExecutionError(`Couldn't find installation for ${account}.`)
  }

  if (DateTime.fromJSDate(installation.periodEndsAt) < DateTime.now()) {
    throw new ExecutionError(`Expired subscription ${installation.account}`)
  }

  return installation
}

/**
 * Returns account configuration related to the provided event.
 */
export async function getAccountConfiguration(ctx: BaseContext): Promise<LSCConfiguration> {
  const installation = await getAccountInstallation(ctx)

  const repo = getLSConfigRepoName(installation.account)
  const ref = `refs/heads/${ctx.payload.repository.default_branch}`

  const rawConfig = await getFile(ctx.octokit, { owner: installation.account, repo, ref }, LS_CONFIG_PATH)

  /* istanbul ignore next */
  if (rawConfig === null) {
    throw new ExecutionError(`No configuration found while loading sources.`)
  }

  const parsedConfig = parseConfig({
    input: rawConfig,
    isPro: installation.plan === 'PAID',
  })

  /* istanbul ignore if */
  if (!parsedConfig.ok) {
    throw new ExecutionError(`Invalid configuration: ${parsedConfig.error}`)
  }

  return parsedConfig.config
}

/**
 * Error that may be safely thrown in the event handler.
 */
export class ExecutionError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ExecutionError'
  }
}

/**
 * Lets you use a context function in a safe environment.
 */
export async function withContext(ctx: BaseContext & { log: pino.Logger }, fn: () => Promise<void>): Promise<void> {
  try {
    await fn()
  } catch (err) /* istanbul ignore next */ {
    if (process.env.NODE_ENV !== 'production') {
      console.error(err)
      return
    }

    if (err instanceof ExecutionError) {
      ctx.log.debug(
        {
          action: ctx.payload.action,
          repo: ctx.payload.repository.name,
          owner: ctx.payload.repository.owner.login,
        },
        err.message,
      )
      // with context and all
      return
    }

    throw err
  }
}
