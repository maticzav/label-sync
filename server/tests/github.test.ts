import nock from 'nock'
import path from 'path'
import { ProbotOctokit } from 'probot'

import { isLabel, isLabelDefinition, bootstrapConfigRepository } from '../src/lib/github'
import { populateTemplate } from '../src/lib/bootstrap'
import { loadTreeFromPath } from '../src/lib/utils'
import { Octokit } from 'server/src/lib/context'

/* Fixtures */
const TEMPLATE_PATH = path.resolve(__dirname, '../../templates/typescript/')
const TEMPLATE = populateTemplate(
  loadTreeFromPath(TEMPLATE_PATH, ['dist', 'node_modules', '.DS_Store', /.*\.log.*/, /.*\.lock.*/]),
  {
    repository: 'config-labelsync',
    repositories: [{ name: 'labelsync' }],
  },
)

describe('github integration:', () => {
  beforeAll(() => {
    nock.disableNetConnect()
  })

  afterAll(() => {
    nock.enableNetConnect()
  })

  let github: Octokit

  beforeEach(() => {
    github = new ProbotOctokit()
  })

  afterEach(() => {
    nock.cleanAll()
  })

  test(
    'bootstraps configuraiton',
    async () => {
      expect.assertions(6)

      // const repoEndpoint = jest.fn().mockReturnValue({})
      const createRepoEndpoint = jest.fn().mockImplementation((uri, body) => {
        expect(body).toEqual({
          name: 'maticzav-labelsync',
          description: 'LabelSync configuration repository.',
          auto_init: true,
        })
        return
      })

      let blobs: { [sha: number]: string } = {}

      const createBlobEndpoint = jest.fn().mockImplementation((uri, body) => {
        const sha = Object.keys(blobs).length
        blobs[sha] = body.content
        return { url: 'url', sha }
      })

      let trees: { [sha: number]: string } = {}

      const createTreeEndpoint = jest.fn().mockImplementation((uri, body) => {
        const sha = Object.keys(trees).length
        trees[sha] = body.tree
        return { sha: sha, url: 'url', tree: body.tree }
      })

      const parentSha = Math.floor(Math.random() * 1000).toString()

      const getRefEndpoint = jest.fn().mockImplementation((uri, body) => {
        return { object: { sha: parentSha } }
      })

      const commitSha = Math.floor(Math.random() * 1000).toString()

      const createCommitEndpoint = jest.fn().mockImplementation((uri, body) => {
        expect(body.parents).toEqual([parentSha])
        return { sha: commitSha }
      })

      const updateRefEndpoint = jest.fn().mockImplementation((uri, body) => {
        expect(body.sha).toBe(commitSha)
        return { object: { sha: '' } }
      })

      /* Mocks */

      // nock('https://api.github.com')
      //   .post('/app/installations/13055/access_tokens')
      //   .reply(200, { token: 'test' })

      // nock('https://api.github.com')
      //   .get('/repos/maticzav/maticzav-labelsync')
      //   .reply(404, repoEndpoint)

      nock('https://api.github.com').post('/orgs/maticzav/repos').reply(200, createRepoEndpoint)

      nock('https://api.github.com')
        .post('/repos/maticzav/maticzav-labelsync/git/blobs')
        .reply(200, createBlobEndpoint)
        .persist()

      nock('https://api.github.com')
        .post('/repos/maticzav/maticzav-labelsync/git/trees')
        .reply(200, createTreeEndpoint)
        .persist()

      nock('https://api.github.com')
        .get('/repos/maticzav/maticzav-labelsync/git/ref/heads%2Fmaster')
        .reply(200, getRefEndpoint)

      nock('https://api.github.com')
        .post('/repos/maticzav/maticzav-labelsync/git/commits')
        .reply(200, createCommitEndpoint)

      nock('https://api.github.com')
        .patch('/repos/maticzav/maticzav-labelsync/git/refs/heads%2Fmaster')
        .reply(200, updateRefEndpoint)

      /* Bootstrap. */

      await bootstrapConfigRepository(github, { owner: 'maticzav', repo: 'maticzav-labelsync' }, TEMPLATE)

      /* Tests */

      expect(createRepoEndpoint).toBeCalledTimes(1)
      expect(blobs).toMatchSnapshot()
      expect(trees).toMatchSnapshot()
    },
    5 * 60 * 1000,
  )
})

describe('github:', () => {
  test('isLabel', () => {
    expect(
      isLabel({
        name: 'bug',
        color: 'ff',
        description: 'desc',
      })({
        name: 'bug',
        description: 'desc',
        color: 'ff',
      }),
    ).toBeTruthy()

    expect(
      isLabel({
        name: 'bug',
        color: '00',
        description: 'desc',
      })({
        name: 'bug',
        description: 'desc',
        color: 'ff',
      }),
    ).toBeFalsy()

    expect(
      isLabel({
        name: 'bug',
        color: '00',
        description: 'desc',
      })({
        name: 'bug/0',
        description: 'this is a bug',
        color: 'ff',
      }),
    ).toBeFalsy()
  })

  test('isLabelDefinition', () => {
    expect(
      isLabelDefinition({
        name: 'bug',
        color: '00',
        description: 'desc',
      })({
        name: 'bug',
        description: 'this is a bug',
        color: 'ff',
      }),
    ).toBeTruthy()

    expect(
      isLabelDefinition({
        name: 'bug',
        color: '00',
        description: 'desc',
      })({
        name: 'bug/0',
        description: 'this is a bug',
        color: 'ff',
      }),
    ).toBeFalsy()
  })
})
