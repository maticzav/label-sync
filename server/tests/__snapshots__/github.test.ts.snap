// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github integration: bootstraps configuraiton 1`] = `
Object {
  "0": "dist
node_modules

.DS_Store
*.log*",
  "1": "# LabelSync TypeScript configuration.

Hey there! Welcome to LabelSync. We have scaffolded the configuration file for you. Check it out!

### Setting up LabelSync

1. Create a repository on Github and name it \`config-labelsync\`.
1. Commit your configuration (this repository) to Github.
1. Head over to [LabelSync Manager Github Application](https://github.com/apps/labelsync-manager) and make sure that you install it in all repositories that you have configured.

### LabelSync library cheat sheet

**Methods:**

- \`labelsync\`: used as a configuration entry point. Outputs yaml version of your configuration to the root of your repository.
- \`repo\`: used to configure a single repository
- \`label\`: used to create a single label

**Presets:**

Check out \`colors\` property with a set of common colors for labels, and \`type\`, \`note\`, \`impact\`, \`effort\`, \`needs\`, \`scope\` and \`communtiy\` label templates to get up to speed more quickly. Label templates all prepend their name to the name of your label and already pack a nice color of our choosing.

\`\`\`ts
function labelsync({
  /* Repositories represent a repo-name:config dictionary */
  repos: { [repo: string]: Repository }
}): Configuration

/* Repo */
function repo({
  config?: {
    /* removes unconfigured labels from repository to keep it clean */
    removeUnconfiguredLabels?: boolean
  }
  /* list of labels that we get using label method below */
  labels: Label[]
})

/* Label */
function label(name: string, color: string)
function label({
  /* name of the label */
  name: string
  /* color in hex format */
  color: string
  description?: string
  /* old names of this label */
  alias?: string[]
  /* siblings of the label */
  siblings?: string[]
})
\`\`\`
",
  "2": "import { labelsync, repo } from \\"label-sync\\";

/* Repository */
import { prisma } from \\"./repos/prisma\\";
import { github } from \\"./repos/github\\";

/* Config */
labelsync({
  repos: {
    /* Check presets in the repos folder. */
    // prisma,
    // github,
    /* Personalized repositories */
    labelsync: repo({
      config: {
        removeUnconfiguredLabels: false,
      },
      labels: [],
    }),
  },
});
",
  "3": "repos: {}
",
  "4": "{
  \\"name\\": \\"config-labelsync\\",
  \\"private\\": true,
  \\"main\\": \\"index.js\\",
  \\"scripts\\": {
    \\"make\\": \\"ts-node labelsync.ts\\"
  },
  \\"dependencies\\": {
    \\"husky\\": \\"latest\\",
    \\"label-sync\\": \\"latest\\"
  },
  \\"devDependencies\\": {
    \\"prettier\\": \\"*\\",
    \\"ts-node\\": \\"*\\",
    \\"typescript\\": \\"*\\"
  },
  \\"husky\\": {
    \\"hooks\\": {
      \\"pre-commit\\": \\"yarn -s make && git add labelsync.yml\\"
    }
  },
  \\"prettier\\": {
    \\"semi\\": false,
    \\"singleQuote\\": true,
    \\"trailingComma\\": \\"all\\",
    \\"printWidth\\": 100
  },
  \\"license\\": \\"MIT\\"
}
",
  "5": "{
  \\"compilerOptions\\": {
    \\"target\\": \\"es2017\\",
    \\"module\\": \\"commonjs\\",
    \\"rootDir\\": \\".\\"
  }
}
",
  "6": "import { repo, label } from \\"label-sync\\";

/**
 * Default collection of label in a Github repository.
 */
export const github = repo({
  config: {
    removeUnconfiguredLabels: false,
  },
  labels: [
    label({
      name: \\"bug\\",
      color: \\"#d73a4a\\",
      description: \\"Something isn't working\\",
    }),
    label({
      name: \\"documentation\\",
      color: \\"#0075ca\\",
      description: \\"Improvements or additions to documentation\\",
    }),
    label({
      name: \\"duplicate\\",
      color: \\"#cfd3d7\\",
      description: \\"This issue or pull request already exists\\",
    }),
    label({
      name: \\"enhancement\\",
      color: \\"#a2eeef\\",
      description: \\"New feature or request\\",
    }),
    label({
      name: \\"good first issue\\",
      color: \\"#7057ff\\",
      description: \\"Good for newcomers\\",
    }),
    label({
      name: \\"help wanted\\",
      color: \\"#008672\\",
      description: \\"Extra attention is needed\\",
    }),
    label({
      name: \\"invalid\\",
      color: \\"#e4e669\\",
      description: \\"This doesn't seem right\\",
    }),
    label({
      name: \\"question\\",
      color: \\"#d876e3\\",
      description: \\"Further information is requested\\",
    }),
    label({
      name: \\"wontfix\\",
      color: \\"#000000\\",
      description: \\"This will not be worked on\\",
    }),
  ],
});
",
  "7": "import { repo, label, colors } from \\"label-sync\\";

/**
 * Label configuration used internally by Prisma team. Labels are grouped
 * by their intention (e.g. bug/*, kind/*, process/*) and give
 * great base for issue triaging.
 */

export const prisma = repo({
  config: {
    removeUnconfiguredLabels: false,
  },
  labels: [
    /* Bugs */
    label({
      name: \\"bug/0-needs-info\\",
      color: colors.danger,
      description: \\"More information is needed for reproduction.\\",
    }),
    label({
      name: \\"bug/1-repro-available\\",
      color: colors.danger,
      description: \\"A reproduction exists and needs to be confirmed.\\",
    }),
    label({
      name: \\"bug/2-confirmed\\",
      color: colors.danger,
      description: \\"We have confirmed that this is a bug.\\",
    }),
    /* Kind */
    label({
      name: \\"kind/bug\\",
      color: colors.neutral,
      description: \\"A reported bug.\\",
    }),
    label({
      name: \\"kind/regression\\",
      color: colors.neutral,
      description: \\"A reported bug in functionality that used to work before.\\",
    }),
    label({
      name: \\"kind/feature\\",
      color: colors.neutral,
      description: \\"A request for a new feature.\\",
    }),
    label({
      name: \\"kind/improvement\\",
      color: colors.neutral,
      description: \\"An improvement to existing feature and code.\\",
    }),
    label({
      name: \\"kind/docs\\",
      color: colors.neutral,
      description: \\"A documentation change is required.\\",
    }),
    label({
      name: \\"kind/discussion\\",
      color: colors.neutral,
      description: \\"Discussion is required.\\",
    }),
    label({
      name: \\"kind/question\\",
      color: colors.neutral,
      description: \\"Developer asked a question. No code changes required.\\",
    }),
    /* Process triaging. */
    label({
      name: \\"process/candidate\\",
      color: colors.shiny,
      description: \\"Candidate for next Milestone.\\",
    }),
    label({
      name: \\"process/next-milestone\\",
      color: colors.shiny,
      description: \\"Issue earmarked for next Milestone.\\",
    }),
    label({
      name: \\"process/product\\",
      color: colors.shiny,
      description:
        \\"Temporary label to export products issues from the Engineering process\\",
    }),
  ],
});
",
}
`;

exports[`github integration: bootstraps configuraiton 2`] = `
Object {
  "0": Array [
    Object {
      "mode": "100644",
      "path": "github.ts",
      "sha": 6,
      "type": "blob",
    },
    Object {
      "mode": "100644",
      "path": "prisma.ts",
      "sha": 7,
      "type": "blob",
    },
  ],
  "1": Array [
    Object {
      "mode": "040000",
      "path": "repos",
      "sha": 0,
      "type": "tree",
    },
    Object {
      "mode": "100644",
      "path": ".gitignore",
      "sha": 0,
      "type": "blob",
    },
    Object {
      "mode": "100644",
      "path": "README.md",
      "sha": 1,
      "type": "blob",
    },
    Object {
      "mode": "100644",
      "path": "labelsync.ts",
      "sha": 2,
      "type": "blob",
    },
    Object {
      "mode": "100644",
      "path": "labelsync.yml",
      "sha": 3,
      "type": "blob",
    },
    Object {
      "mode": "100644",
      "path": "package.json",
      "sha": 4,
      "type": "blob",
    },
    Object {
      "mode": "100644",
      "path": "tsconfig.json",
      "sha": 5,
      "type": "blob",
    },
  ],
}
`;
